vllm/inputs/data.py

```diff
@@ -92,15 +94,106 @@ class TokensPrompt(TypedDict):
+prompts explicitly, i.e. ExplicitEncoderDecoderPrompt
+
+A prompt of type SingletonPromptInputs may be employed
+as (1) input to a decoder-only model, (2) input to
+the encoder of an encoder/decoder model, in the scenario
+where the decoder-prompt is not specified explicitly, or
+(3) as a member of a larger data structure encapsulating
+more than one prompt, i.e. ExplicitEncoderDecoderPrompt
+"""
+
+
+class ExplicitEncoderDecoderPrompt(TypedDict):
+    """Represents an encoder/decoder model input prompt,
+    comprising an explicit encoder prompt and a 
+    decoder prompt.
+
+    The encoder and decoder prompts, respectively,
+    may formatted according to any of the
+    SingletonPromptInputs schemas, and are not
+    required to have the same schema.
+
+    Only the encoder prompt may have multi-modal data.
+
+    Note that an ExplicitEncoderDecoderPrompt may not
+    be used as an input to a decoder-only model,
+    and that the `encoder_prompt` and `decoder_prompt`
+    fields of this data structure may not themselves
+    must be SingletonPromptInputs instances.
+    """
+
+    encoder_prompt: SingletonPromptInputs
+
+    decoder_prompt: SingletonPromptInputs
+
+
+PromptInputs = Union[SingletonPromptInputs, ExplicitEncoderDecoderPrompt]
 """
-The inputs to the LLM, which can take one of the following forms:
+Set of possible schemas for an LLM input, including
+both decoder-only and encoder/decoder input types:
 
 - A text prompt (:class:`str` or :class:`TextPrompt`)
 - A tokenized prompt (:class:`TokensPrompt`)
+- A single data structure containing both an encoder and a decoder prompt
+  (:class:`ExplicitEncoderDecoderPrompt`)
 """
 
 
+def get_prompt_type(prompt: Optional[PromptInputs], ) -> Optional[str]:
+    """
+    Get the type-name of the prompt argument instance, given that
+    isinstance() cannot apply to TypedDict subclasses directly.
+    If the prompt is None, return 'None' as the type name.
+
+    Arguments:
+
+    * prompt: LLM input prompt or None
+
+    Returns:
+
+    * String representation of prompt type
+    """
+
+    if prompt is None:
+        return 'None'
+
+    required_keys_dict = {
+        'TextPrompt': ['prompt'],
+        'TokensPrompt': ['prompt_token_ids'],
+        'ExplicitEncoderDecoder': ['encoder_prompt', 'decoder_prompt'],
+    }
+
+    if is_dict(prompt):
+        for ptype in required_keys_dict:
+            # Ignore type checking in the conditional below because type
+            # checker does not understand that is_dict(prompt) narrows
+            # down the possible types
+            if has_required_keys(
+                    prompt,  # type: ignore
+                    required_keys_dict[ptype]):
+                return ptype
+
+        raise ValueError(f"Invalid prompt {prompt}, valid types are "
+                         "required_keys_dict={required_keys_dict}")
+    elif is_str(prompt):
+        return "str"
+
+    raise ValueError(f"Invalid prompt {prompt}")
```

Comment:
IMO it would be easier to just use pydantic's `TypeAdapter` to check the prompt format rather than implementing our own.

Reply:
Hi @DarkLight1337 I appreciate your feedback and I spent the last 30min or so attempting to re-implement this function based on Pydantic for type validation. Unfortunately after having done so, I think that keeping the function as-is makes more sense (I have implemented all of Nick's feedback on the function.)

The function you are referring to, `get_prompt_type()` is for checking prompt types; the prompt types defined in vllm/inputs/data.py are all `TypedDict` subclasses.

I appreciate that pydantic is in principle the right tool for the job here. Unfortunately since vLLM currently advertises itself as supporting Python 3.8 - 3.12, we run into a compatibility problem between TypedDict and Pydantic:

https://docs.pydantic.dev/2.3/usage/types/dicts_mapping/#typeddict

The following note is at the link above:

"This {typing.TypedDict} is a new feature of the Python standard library as of Python 3.8. Because of limitations in typing.TypedDict before 3.12, the [typing-extensions](https://pypi.org/project/typing-extensions/) package is required for Python <3.12. You'll need to import TypedDict from typing_extensions instead of typing and will get a build time error if you don't."

I think the widespread use of typing.TypedDict throughout the codebase makes it problematic to try and use typing_extensions.TypedDict in just a few places, for compatibility reasons.

In fact, I tried re-implementing the function as you suggested, making sure to import typing_extensions.TypedDict in vllm/inputs/data.py and use it in all places in that file instead of typing.TypedDict. The re-implemented function is here:

```
def get_prompt_type(prompt: Optional[PromptInputs], ) -> Optional[str]:
    """
    ...
    """

    if prompt is None:
        return 'None'

    if isinstance(prompt, str):
        return "str"

    # If the prompt type is valid, it must be a dict
    # by process of elimination
    text_prompt_adapter = TypeAdapter(TextPrompt)
    if text_prompt_adapter.validate_python(prompt):
        return 'TextPrompt'
    tokens_prompt_adapter = TypeAdapter(TokensPrompt)
    if tokens_prompt_adapter.validate_python(prompt):
        return 'TokensPrompt'
    explicit_encoder_decoder_prompt_adapter = (
        TypeAdapter(ExplicitEncoderDecoderPrompt)
        )
    if explicit_encoder_decoder_prompt_adapter.validate_python(prompt):
        return 'ExplicitEncoderDecoder'

    raise ValueError(f"Invalid prompt {prompt}")
```

What I find is that, even though all of the prompt types - TextPrompt, TokensPrompt, etc. - are defined using typing_extensions.TypedDict in my new implementation, nonetheless the line `text_prompt_adapter = TypeAdapter(TextPrompt)` fails with an error message that I should be using typing_extensions.TypedDict instead of typing.TypedDict. Close inspection has not allowed me to figure out the issue causing this, but I suspect it has to do with the use of typing.TypedDict in the broader codebase.

Anyway, my larger point is - I think the effort of using pydantic here may exceed the benefit. So I plan to keep this function implementation the way it is. Perhaps it would be easier to revisit this in the future once the this typing_extensions compatibility issue is no longer relevant. However please feel free to restart this conversation if you strongly disagree with keeping this function as-is.

Reply:
In that case, it's fine to use our current implementation. Thanks for trying this out regardless!

Reply:
Actually I just got another idea: how about using similar logic as `parse_and_batch_prompt` to return tagged unions (a dictionary with a key used to discriminate between the results, and the original input but cast to the correct data type). For example:

```py
class ParsedTextPrompt(TypedDict):
    type: Literal["TextPrompt"]
    prompt: TextPrompt

class ParsedTokensPrompt(TypedDict):
    type: Literal["TokensPrompt"]
    prompt: TokensPrompt

class ParsedExplicitEncoderDecoderPrompt(TypedDict):
    type: Literal["ExplicitEncoderDecoderPrompt"]
    prompt: ExplicitEncoderDecoderPrompt

PromptDict = Union[TextPrompt, TokensPrompt, ExplicitEncoderDecoderPrompt]
ParsedPromptDict = Union[ParsedTextPrompt, ParsedTokensPrompt, ParsedExplicitEncoderDecoderPrompt]

def parse_prompt_dict_type(prompt: PromptDict) -> ParsedPromptDict:
    ...
```

Then in downstream usage:
```py
parsed_prompt_dict = parse_prompt_dict_type(prompt)
if parsed_prompt_dict["type"] == "TextPrompt":
    prompt_dict = parsed_prompt_dict["prompt"]  # Automatically inferred as TextPrompt
    ...  # Processing logic
elif parsed_prompt_dict["type"] == "TokensPrompt":
    prompt_dict = parsed_prompt_dict["prompt"]  # Automatically inferred as TokensPrompt
    ...  # Processing logic
elif parsed_prompt_dict["type"] == "ExplicitEncoderDecoderPrompt":
    prompt_dict = parsed_prompt_dict["prompt"]  # Automatically inferred as ExplicitEncoderDecoderPrompt
    ...  # Processing logic
```

Reply:
Hi @DarkLight1337 good idea, thanks for the suggestion. quick update - I am working on implementing this. If I can get it to work without issue then I'll incorporate it into this PR.

If I run into any non-obvious issues doing this approach, I might end up pushing this change to a future PR

Reply:
I have created a new PR #7258 to address this.